docs/multi-stage-query/reference.md

```diff
@@ -88,93 +118,260 @@ FROM TABLE(
+objects to read:
 
-| Name | Description | JSON equivalent | Required |
-| ---- | ----------- | --------------- | -------- |
-| `listDelimiter` | The delimiter to use for fields that represent a list of strings. | `listDelimiter` | No |
-| `skipRows` | The number of rows to skip at the start of the file. Default is 0. | `skipHeaderRows` | No |
+* `uris` (`ARRAY` of `VARCHAR`)
+* `prefix` (`VARCHAR`) - Corresponds to the JSON `prefixes` property, but allows a single
+  prefix.
+* `bucket` (`VARCHAR`) - Corresponds to the `bucket` field of the `objects` JSON field. SQL
+  does not have syntax for an array of objects. Instead, this function taks a single bucket,
+  and one or more objects within that bucket.
+* `paths` (`ARRAY` of `VARCHAR`) - Corresponds to the `path` fields of the `object` JSON field.
+  All paths are within the single `bucket` parameter.
 
-MSQ does not have the ability to infer schema from a CSV, file, so the `findColumnsFromHeader` property
-is unavailable. Instead, Columns are given using the `EXTEND` syntax described above.
+The S3 input source accepts one of the following patterns:
 
-#### Delimited Text Format
+* `uris` - A list of fully-qualified object URIs.
+* `prefixes` - A list of fully-qualified "folder" prefixes.
+* `bucket` and `paths` - A list of objects relative to the given bucket path.
+
+The `S3` function also accepts the following security properties:
 
-Use the `tsv` format to read from an arbitrary delimited (CSV-like) file such as tab-delimited,
-pipe-delimited, etc. This choice selects the Druid `DelimitedInputFormat` class.
+* `accessKeyId` (`VARCHAR`)
+* `secretAccessKey` (`VARCHAR`)
+* `assumeRoleArn` (`VARCHAR`)
 
-| Name | Description | JSON equivalent | Required |
-| ---- | ----------- | --------------- | -------- |
-| `delimiter` | The delimiter which separates fields. | `delimiter` | Yes |
-| `listDelimiter` | The delimiter to use for fields that represent a list of strings. | `listDelimiter` | No |
-| `skipRows` | The number of rows to skip at the start of the file. Default is 0. | `skipHeaderRows` | No |
+The `S3` table function does not support either the `clientConfig` or `proxyConfig`
+JSON properties.
 
-As noted above, MSQ cannot infer schema using headers. Use `EXTEND` instead.
+If you need the full power of the S3 input source, then consider the use of the `extern`
+function, which accepts the full S3 input source serializd as JSON. Alternatively,
+create a catalog external table that has the full set of properties, leaving just the
+`uris` or `paths` to be provided at query time.
+
+Examples, each of which correspond to an example on the
+[S3 input source](../ingestion/native-batch-input-sources.md#s3-input-source) page.
+The examples omit the format and schema; however you must remember to provide those
+in an actual query.
+
+```sql
+  TABLE(S3(uris => ARRAY['s3://foo/bar/file.json', 's3://bar/foo/file2.json']))
```

Comment:
nit: to make consistent with the format of the localFiles function given above, should we drop the `TABLE(` prefix here, and specify the format?